commit 9e49ca756d207f4313fb7af48648a67da8e4e250
Author: Steven Rostedt <rostedt@goodmis.org>
Date:   Fri Dec 20 10:33:13 2024 -0500

    tracing/string: Create and use __free(argv_free) in trace_dynevent.c
    
    The function dyn_event_release() uses argv_split() which must be freed via
    argv_free(). It contains several error paths that do a goto out to call
    argv_free() for cleanup. This makes the code complex and error prone.
    
    Create a new __free() directive __free(argv_free) that will call
    argv_free() for data allocated with argv_split(), and use it in the
    dyn_event_release() function.
    
    Cc: Kees Cook <kees@kernel.org>
    Cc: Masami Hiramatsu <mhiramat@kernel.org>
    Cc: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
    Cc: Peter Zijlstra <peterz@infradead.org>
    Cc: Andy Shevchenko <andy@kernel.org>
    Cc: linux-hardening@vger.kernel.org
    Link: https://lore.kernel.org/20241220103313.4a74ec8e@gandalf.local.home
    Signed-off-by: Steven Rostedt (Google) <rostedt@goodmis.org>

diff --git a/include/linux/string.h b/include/linux/string.h
index 493ac4862c77..86d5d352068b 100644
--- a/include/linux/string.h
+++ b/include/linux/string.h
@@ -4,6 +4,7 @@
 
 #include <linux/args.h>
 #include <linux/array_size.h>
+#include <linux/cleanup.h>	/* for DEFINE_FREE() */
 #include <linux/compiler.h>	/* for inline */
 #include <linux/types.h>	/* for size_t */
 #include <linux/stddef.h>	/* for NULL */
@@ -312,6 +313,8 @@ extern void *kmemdup_array(const void *src, size_t count, size_t element_size, g
 extern char **argv_split(gfp_t gfp, const char *str, int *argcp);
 extern void argv_free(char **argv);
 
+DEFINE_FREE(argv_free, char **, if (!IS_ERR_OR_NULL(_T)) argv_free(_T))
+
 /* lib/cmdline.c */
 extern int get_option(char **str, int *pint);
 extern char *get_options(const char *str, int nints, int *ints);